# Loan-Default-Analysis

## 資料集說明（Dataset Description） 
**貸款違約資料集包含從2013至2017年，有關已申請貸款的借款人的信息，以及有關其財務狀況、貸款特徵和還款行為的詳細資訊:** 


LoanID : A unique identifier for each loan in the dataset.

Age : The borrower's age at the time the loan was issued.

Income : The borrower's annual income

LoanAmount : The total amount of the loan that the borrower is requesting or has been approved for.

CreditScore : A numerical representation of the borrower's creditworthiness, typically ranging from 300 to 850. A higher credit score indicates the borrower is more likely to repay the debt.

MonthsEmployed : The number of months the borrower has been employed at their current job or with their current employer.

NumCreditLines : The total number of active credit lines (e.g., credit cards, loans) the borrower has at the time of applying for the loan.

InterestRate : The annual percentage rate (APR) charged for borrowing the loan amount, usually expressed as a percentage.

LoanTerm : The length of time (in months) over which the loan is to be repaid.

DTIRatio : The Debt-to-Income ratio, which measures the borrower’debt payments relative to their income. A higher ratio can indicate greater financial stress.

Education : The highest level of education the borrower has completed (e.g., High School, Bachelor’s, Master’s, etc.).

EmploymentType : The type of employment the borrower is engaged in (e.g., Full-Time, Part-Time, Self-Employed, etc.).

MaritalStatus : The marital status of the borrower (e.g., Single, Married, Divorced, etc.).

HasMortgage : An indicator (e.g., Yes/No) that shows whether the borrower has an existing mortgage on a property.

HasDependents : An indicator (e.g., Yes/No) that shows whether the borrower has dependents (children, other family members) to support.

LoanPurpose : The primary reason for taking out the loan (e.g., Home Purchase, Debt Consolidation, Education, etc.).

HasCoSigner : An indicator (e.g., Yes/No) that shows whether the borrower has a co-signer for the loan (someone who agrees to take responsibility if the borrower defaults).

Default : An indicator (e.g., Yes/No) that shows whether the borrower defaulted on the loan or failed to make timely payments.

Loan Date (DD/MM/YYYY) : The date the loan was issued or originated.

## 1.讀取數據

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv(r"C:\Users\User\Desktop\PowerBI\Data\Loan_default.csv")
original_data.head()

LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0  I38PQUQS96   56   85994       50587          520              80   
1  HPSK72WA7R   69   50432      124440          458              15   
2  C1OZ6DPJ8Y   46   84208      129188          451              26   
3  V2KKSFM3UN   32   31713       44799          743               0   
4  EY08JDHTZP   60   20437        9139          633               8   

   NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
0               4         15.23        36      0.44   Bachelor's   
1               1          4.81        60      0.68     Master's   
2               3         21.17        24      0.31     Master's   
3               3          7.07        24      0.23  High School   
4               4          6.51        48      0.73   Bachelor's   

  EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
0      Full-time      Divorced         Yes           Yes       Other   
1      Full-time       Married          No            No       Other   
2     Unemployed      Divorced         Yes           Yes        Auto   
3      Full-time       Married          No            No    Business   
4     Unemployed      Divorced          No           Yes        Auto   

  HasCoSigner  Default Loan Date (DD/MM/YYYY)  
0         Yes        0             10/15/2018  
1         Yes        0              3/25/2016  
2          No        1             11/11/2013  
3          No        0              6/22/2017  
4          No        0               6/9/2014

## 2.評估數據

**2.1 評估數據整齊度**  
為了區分開經過清理的資料和原始資料，我們創建新的變數`cleaned_data`讓它為`original_data`複製的副本。

In [3]:
cleaned_data = original_data.copy()
cleaned_data.sample(10)

LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
216022  HWRG1OGU3U   28  110153       70862          574              29   
138324  3S07P9NGRX   50  120017      206913          747              47   
59248   4WRVK0VDAR   22   46505        8143          645              33   
76970   4A05G3ARFN   32  108132      236457          807              12   
50961   BJDCKMEJ4N   29   66812       25657          837             102   
234161  4OHM5N7DQO   50   75784       75285          469               8   
190230  28KP84C6LY   35   74666      171622          837             113   
96079   AA1Y5QPKMU   63   26387       85408          728              89   
34302   PXW5H1IJR4   46   96665      214237          825              93   
119057  RIABICTCXB   58  147745      224452          323              22   

        NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
216022               2          3.38        36      0.15  High School   
138324               4         21.44        60      0.32   Bachelor's   
59248                2          5.28        60      0.53  High School   
76970                4          6.62        48      0.87     Master's   
50961                1         21.69        12      0.42  High School   
234161               2          4.03        60      0.76   Bachelor's   
190230               4         19.89        60      0.88          PhD   
96079                2         18.55        12      0.60     Master's   
34302                2         11.09        60      0.88          PhD   
119057               1         17.10        24      0.85  High School   

       EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
216022  Self-employed      Divorced         Yes            No    Business   
138324      Part-time       Married          No            No       Other   
59248       Part-time        Single          No           Yes        Home   
76970       Full-time        Single         Yes            No        Home   
50961      Unemployed        Single         Yes            No        Auto   
234161      Full-time        Single          No            No        Auto   
190230  Self-employed      Divorced         Yes            No   Education   
96079   Self-employed       Married         Yes           Yes        Auto   
34302       Part-time      Divorced          No           Yes   Education   
119057      Full-time       Married          No           Yes        Auto   

       HasCoSigner  Default Loan Date (DD/MM/YYYY)  
216022         Yes        0             10/19/2018  
138324         Yes        0              8/22/2017  
59248           No        0              1/20/2016  
76970          Yes        0              6/28/2016  
50961           No        1               1/7/2015  
234161          No        0              6/19/2016  
190230          No        1              5/28/2014  
96079          Yes        0              7/21/2016  
34302           No        0              8/20/2013  
119057          No        0             12/12/2014

我們發現`cleaned_data`符合數據整齊度的要求("每列一個變量，每行一個觀察值，每個單元格是一個值")

## 3.數據乾淨度

**3.1 對`cleaned_data`進行調整**

In [4]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255347 entries, 0 to 255346
Data columns (total 19 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   LoanID                  255347 non-null  object 
 1   Age                     255347 non-null  int64  
 2   Income                  255347 non-null  int64  
 3   LoanAmount              255347 non-null  int64  
 4   CreditScore             255347 non-null  int64  
 5   MonthsEmployed          255347 non-null  int64  
 6   NumCreditLines          255347 non-null  int64  
 7   InterestRate            255347 non-null  float64
 8   LoanTerm                255347 non-null  int64  
 9   DTIRatio                255347 non-null  float64
 10  Education               255347 non-null  object 
 11  EmploymentType          255347 non-null  object 
 12  MaritalStatus           255347 non-null  object 
 13  HasMortgage             255347 non-null  object 
 14  HasDependents       

從輸出結果，`cleaned_data`資料共有255347筆觀察值，並且沒有發現缺失值。  
我們將`Loan Date (DD/MM/YYYY)`轉換為日期。

In [5]:
cleaned_data["Loan Date (DD/MM/YYYY)"] = pd.to_datetime(cleaned_data["Loan Date (DD/MM/YYYY)"], format = '%m/%d/%Y')
cleaned_data["Loan Date (DD/MM/YYYY)"]

0        2018-10-15
1        2016-03-25
2        2013-11-11
3        2017-06-22
4        2014-06-09
            ...    
255342   2017-10-07
255343   2016-01-01
255344   2017-03-05
255345   2013-04-05
255346   2015-01-27
Name: Loan Date (DD/MM/YYYY), Length: 255347, dtype: datetime64[ns]

**3.2 處理重複數據**  
`LoanID`裡不應該存在相同的觀察值，因此查看是否存在重複值

In [6]:
cleaned_data["LoanID"].duplicated().sum()

np.int64(0)

結果是0，代表不存在重複值

**3.3處理不一致數據**  
不一致資料可能在於`Education`,`EmploymentType`,`MaritalStatus`,`HasMortgage`,`HasDependents`,`LoanPurpose	`和`HasCoSigner`

In [7]:
cleaned_data["Education"].value_counts()

Education
Bachelor's     64366
High School    63903
Master's       63541
PhD            63537
Name: count, dtype: int64

In [8]:
cleaned_data["EmploymentType"].value_counts()

EmploymentType
Part-time        64161
Unemployed       63824
Self-employed    63706
Full-time        63656
Name: count, dtype: int64

In [9]:
cleaned_data["MaritalStatus"].value_counts()

MaritalStatus
Married     85302
Divorced    85033
Single      85012
Name: count, dtype: int64

In [10]:
cleaned_data["HasMortgage"].value_counts()

HasMortgage
Yes    127677
No     127670
Name: count, dtype: int64

In [11]:
cleaned_data["HasDependents"].value_counts()

HasDependents
Yes    127742
No     127605
Name: count, dtype: int64

In [12]:
cleaned_data["LoanPurpose"].value_counts()

LoanPurpose
Business     51298
Home         51286
Education    51005
Other        50914
Auto         50844
Name: count, dtype: int64

In [13]:
cleaned_data["HasCoSigner"].value_counts()

HasCoSigner
Yes    127701
No     127646
Name: count, dtype: int64

結果發現，並不存在不一致資料。

**3.4處理無效或錯誤數據**

In [14]:
cleaned_data.describe()

Age         Income     LoanAmount    CreditScore  \
count  255347.000000  255347.000000  255347.000000  255347.000000   
mean       43.498306   82499.304597  127578.865512     574.264346   
min        18.000000   15000.000000    5000.000000     300.000000   
25%        31.000000   48825.500000   66156.000000     437.000000   
50%        43.000000   82466.000000  127556.000000     574.000000   
75%        56.000000  116219.000000  188985.000000     712.000000   
max        69.000000  149999.000000  249999.000000     849.000000   
std        14.990258   38963.013729   70840.706142     158.903867   

       MonthsEmployed  NumCreditLines   InterestRate       LoanTerm  \
count   255347.000000   255347.000000  255347.000000  255347.000000   
mean        59.541976        2.501036      13.492773      36.025894   
min          0.000000        1.000000       2.000000      12.000000   
25%         30.000000        2.000000       7.770000      24.000000   
50%         60.000000        2.000000      13.460000      36.000000   
75%         90.000000        3.000000      19.250000      48.000000   
max        119.000000        4.000000      25.000000      60.000000   
std         34.643376        1.117018       6.636443      16.969330   

            DTIRatio        Default         Loan Date (DD/MM/YYYY)  
count  255347.000000  255347.000000                         255347  
mean        0.500212       0.116128  2016-01-01 19:16:36.916353024  
min         0.100000       0.000000            2013-01-01 00:00:00  
25%         0.300000       0.000000            2014-07-03 00:00:00  
50%         0.500000       0.000000            2016-01-02 00:00:00  
75%         0.700000       0.000000            2017-07-04 00:00:00  
max         0.900000       1.000000            2018-12-31 00:00:00  
std         0.230917       0.320379                            NaN

從以上統計資訊來看，不存在脫離現實意義的數據

## 4.保存清理後的數據並匯入MySQL

In [19]:
cleaned_data.to_csv(r"C:\Users\User\Desktop\python\Loan_cleaned.csv", index=False)

In [20]:
pd.read_csv(r"C:\Users\User\Desktop\python\Loan_cleaned.csv").head()

LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0  I38PQUQS96   56   85994       50587          520              80   
1  HPSK72WA7R   69   50432      124440          458              15   
2  C1OZ6DPJ8Y   46   84208      129188          451              26   
3  V2KKSFM3UN   32   31713       44799          743               0   
4  EY08JDHTZP   60   20437        9139          633               8   

   NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
0               4         15.23        36      0.44   Bachelor's   
1               1          4.81        60      0.68     Master's   
2               3         21.17        24      0.31     Master's   
3               3          7.07        24      0.23  High School   
4               4          6.51        48      0.73   Bachelor's   

  EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
0      Full-time      Divorced         Yes           Yes       Other   
1      Full-time       Married          No            No       Other   
2     Unemployed      Divorced         Yes           Yes        Auto   
3      Full-time       Married          No            No    Business   
4     Unemployed      Divorced          No           Yes        Auto   

  HasCoSigner  Default Loan Date (DD/MM/YYYY)  
0         Yes        0             2018-10-15  
1         Yes        0             2016-03-25  
2          No        1             2013-11-11  
3          No        0             2017-06-22  
4          No        0             2014-06-09

### 已完成對資料集`cleaned_data`進行資料清理和整理，接下來把`cleaned_data`的資料集匯入MySQL進行**資料轉換**。

**1.安裝與載入必要套件**

In [15]:
pip install sqlalchemy pymysql

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [16]:
import pandas as pd
from sqlalchemy import create_engine

**2.建立 MySQL 連線**(已建立好資料庫`loan_db`)

In [17]:
engine = create_engine(
    "mysql+pymysql://root:derekun98@localhost:3306/loan_db"
)
##"mysql+pymysql://使用者名稱:密碼@localhost:3306/資料庫名稱"

**3.將`cleaned_data`匯入MySQL**

In [18]:
cleaned_data.to_sql(
    name="loan_cleaned",
    con=engine,
    if_exists="replace",   # replace / append
    index=False
)

RuntimeError: 'cryptography' package is required for sha256_password or caching_sha2_password auth methods